### PR TITLE
Detect and format timestamp attributes

### PR DESCRIPTION
**Pull request by spacegaier**
## Proposed change

Ensures that timestamps in more-info popus for attributes are properly formatted.

One point to discuss:
- [ ] <s>The RegExp I use also matches pure dates without a time element (but not the other way around) => should I add a second RegExp that then checks if it is a pure date and then we call `formatDate` rather than `formatDateTimeWithSeconds` as I do currently?</s>
- [X] Adjust the RegExp to not match dates, which solves the point above.
- [X] Added 2nd RegExp for date detection and formatting.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [X] The code change is tested and works locally.
- [X] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/common/string/is_timestamp.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export const isTimestamp = (input: string): boolean => {
+  // https://stackoverflow.com/a/14322189/1947205
+  const regexp = /^([+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24:?00)([.,]\d+(?!:))?)?(\17[0-5]\d([.,]\d+)?)?([zZ]|([+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$/;
```
Comment (bramkragten) [inline src/common/string/is_timestamp.ts:3]: Move this out of the function, no need to recompile the regex everytime

File: src/components/ha-attributes.ts [hunk 2]
```diff
@@ -116,19 +125,38 @@ class HaAttributes extends LitElement {
...
+      // Timestamp handling
+      if (isTimestamp(value)) {
+        const date = new Date(value);
+        if (checkValidDate(date)) {
+          return formatDateTimeWithSeconds(date, this.hass.locale);
+        }
+      }
+
+      // Date handling
+      if (isDate(value)) {
```
Comment (spacegaier) [inline src/components/ha-attributes.ts:152]: Could be potentially optimized by first checking if the string starts with a date and if that is the case and there are more characters left after it then validate against a timestamp. But not sure if worth the effort and complexity.

Comment (bramkragten) [reply src/components/ha-attributes.ts:152]: If `isDate` also matches a timestamp, you should move the timestamp check in here.

Comment (spacegaier) [reply src/components/ha-attributes.ts]: I adjusted the RegExp and their order for the discussed runtime efficiency.

File: src/common/string/is_date.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+const regExpString = "/^\\d{4}-(0[1-9]|1[0-2])-([12]\\d|0[1-9]|3[01])";
+
+const regExp = new RegExp(regExpString + "$/");
+// 2nd expression without the "end of string" enforced, so it can be used
+// to just verify the start of a string and then based on that result e.g.
+// check for a full timestamp string efficiently.
+const regExpNoStringEnd = new RegExp(regExpString + "/");
+
+export const isDate = (input: string, allowCharsAfterDate = false): boolean =>
+  allowCharsAfterDate ? regExp.test(input) : regExpNoStringEnd.test(input);
```
Comment (bramkragten) [inline src/common/string/is_date.ts:11]: this is the wrong way?
```suggestion
  allowCharsAfterDate ? regExpNoStringEnd.test(input) : regExp.test(input);
```

File: src/panels/lovelace/special-rows/hui-attribute-row.ts [hunk 4]
```diff
@@ -72,7 +73,7 @@ class HuiAttributeRow extends LitElement implements LovelaceRow {
               ></hui-timestamp-display>`
             : typeof attribute === "number"
             ? formatNumber(attribute, this.hass.locale)
-            : attribute ?? "-"}
+            : formatAttributeValue(this.hass, attribute ?? "-")}
```
Comment (bramkragten) [inline src/panels/lovelace/special-rows/hui-attribute-row.ts:76]: No need to run those regex on the "-"
```suggestion
            : attribute ? formatAttributeValue(this.hass, attribute) : "-"}
```

Comment (bramkragten) [reply src/panels/lovelace/special-rows/hui-attribute-row.ts:76]: This should be changed in the other places too.

**Review by bramkragten — APPROVED** [review]

File: src/panels/lovelace/cards/hui-entity-card.ts [hunk 5]
```diff
@@ -124,8 +125,11 @@ export class HuiEntityCard extends LitElement implements LovelaceCard {
         <div class="info">
           <span class="value"
             >${"attribute" in this._config
-              ? stateObj.attributes[this._config.attribute!] ??
-                this.hass.localize("state.default.unknown")
+              ? formatAttributeValue(
+                  this.hass,
+                  stateObj.attributes[this._config.attribute!] ??
+                    this.hass.localize("state.default.unknown")
+                )
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-entity-card.ts:132]: We should not call `formatAttributeValue` on `this.hass.localize("state.default.unknown")`

Comment (spacegaier) [reply src/panels/lovelace/cards/hui-entity-card.ts:132]: https://github.com/home-assistant/frontend/pull/9252

**Pull request merged**